src/replication.c

```diff
@@ -3740,6 +3740,15 @@ void replicationSetPrimary(char *ip, int port, int full_sync_required) {
     sdsfree(server.primary_host);
     server.primary_host = NULL;
     if (server.primary) {
+        if (full_sync_required) {
+            /* If full sync is required, we add the dont_cache_primary flag.
+             * So in freeClient (or in freeClientAsync), we won't cache the
+             * primary client. Doing so increases this replica node's election
+             * rank (delay) and reduces its chance of winning the election.
+             * If a replica requiring a full sync wins the election, it will
+             * flush valid data in the shard, causing data loss. */
+            server.primary->flag.dont_cache_primary = 1;
+        }
```

Comment:
```suggestion
    /* When joining 'myself' to a new primary, set the dont_cache_primary flag
     * if a full sync is required. This happens when 'myself' was previously
     * part of a different shard from the new primary. Since 'myself' does not
     * have the replication history of the shard it is joining, clearing the 
     * cached primary is necessary to ensure proper replication behavior. */
    server.primary->flag.dont_cache_primary = full_sync_required;
```